flink/flink-1.14/src/main/java/io/tidb/bigdata/flink/connector/sink/output/InsertOnDuplicateKeyUpdateOutputFormatBuilder.java

```diff
@@ -0,0 +1,175 @@
+import org.apache.flink.connector.jdbc.internal.options.JdbcConnectorOptions;
+import org.apache.flink.connector.jdbc.internal.options.JdbcDmlOptions;
+import org.apache.flink.connector.jdbc.statement.FieldNamedPreparedStatement;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.RowType;
+
+/** Builder for {@link JdbcOutputFormat} for clause `INSERT ... ON DUPLICATE KEY UPDATE`. */
+public class InsertOnDuplicateKeyUpdateOutputFormatBuilder implements Serializable {
+
+  private static final long serialVersionUID = 1L;
+
+  private JdbcConnectorOptions jdbcOptions;
+  private JdbcExecutionOptions executionOptions;
+  private JdbcDmlOptions dmlOptions;
+  private String[] updateColumnNames;
+  private DataType[] updateColumnTypes;
+  private int[] updateColumnIndexes;
+
+  public InsertOnDuplicateKeyUpdateOutputFormatBuilder() {}
+
+  public InsertOnDuplicateKeyUpdateOutputFormatBuilder setJdbcOptions(
+      JdbcConnectorOptions jdbcOptions) {
+    this.jdbcOptions = jdbcOptions;
+    return this;
+  }
+
+  public InsertOnDuplicateKeyUpdateOutputFormatBuilder setJdbcExecutionOptions(
+      JdbcExecutionOptions executionOptions) {
+    this.executionOptions = executionOptions;
+    return this;
+  }
+
+  public InsertOnDuplicateKeyUpdateOutputFormatBuilder setJdbcDmlOptions(
+      JdbcDmlOptions dmlOptions) {
+    this.dmlOptions = dmlOptions;
+    return this;
+  }
+
+  public InsertOnDuplicateKeyUpdateOutputFormatBuilder setUpdateColumnNames(
+      String[] updateColumnNames) {
+    this.updateColumnNames = updateColumnNames;
+    return this;
+  }
+
+  public InsertOnDuplicateKeyUpdateOutputFormatBuilder setUpdateColumnTypes(
+      DataType[] updateColumnTypes) {
+    this.updateColumnTypes = updateColumnTypes;
+    return this;
+  }
+
+  public InsertOnDuplicateKeyUpdateOutputFormatBuilder setUpdateColumnIndexes(
+      int[] updateColumnIndexes) {
+    this.updateColumnIndexes = updateColumnIndexes;
+    return this;
+  }
+
+  public JdbcOutputFormat<RowData, ?, ?> build() {
+    checkNotNull(jdbcOptions, "jdbc options can not be null");
+    checkNotNull(dmlOptions, "jdbc dml options can not be null");
+    checkNotNull(executionOptions, "jdbc execution options can not be null");
+    checkNotNull(updateColumnNames, "updateColumnNames options can not be null");
+    checkNotNull(updateColumnTypes, "updateColumnTypes options can not be null");
+    checkNotNull(updateColumnIndexes, "updateColumnIndexes options can not be null");
+
+    final LogicalType[] logicalTypes =
+        Arrays.stream(updateColumnTypes).map(DataType::getLogicalType).toArray(LogicalType[]::new);
+
+    // upsert query
+    return new JdbcOutputFormat<>(
+        new SimpleJdbcConnectionProvider(jdbcOptions),
+        executionOptions,
+        ctx ->
+            createBufferReduceExecutor(
```

Comment:
it is better to use createSimpleBufferedExecutor?

Reply:
Why?

Reply:
you will call createSimpleRowExecutor, and flink jdbc will use createSimpleBufferedExecutor to call createSimpleRowExecutor

Reply:
https://github.com/tidb-incubator/TiBigData/blob/9fb74a7d964c2ed317430c41b8094a4cfa2153c3/flink/flink-1.14/src/main/java/io/tidb/bigdata/flink/connector/sink/output/InsertOnDuplicateKeyUpdateOutputFormatBuilder.java#L131-L146

Reply:
It indeed return `TableBufferedStatementExecutor`.

Reply:
SimpleRowExecutor is a constructor parameter of TableBufferedStatementExecutor

Reply:
yes. but in Flink JDBC
- createBufferReduceExecutor will return TableBufferReducedStatementExecutor
- createSimpleBufferedExecutor will return TableBufferedStatementExecutor
we'd better meet this rule?

https://github.com/apache/flink/blob/412b49a0fa4756bc6a663782eec3252411bcbd5e/flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcOutputFormatBuilder.java#L129

https://github.com/apache/flink/blob/412b49a0fa4756bc6a663782eec3252411bcbd5e/flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcOutputFormatBuilder.java#L165

Reply:
LGTM